crates/graphql_artifact_generation/src/format_parameter_type.rs

```diff
@@ -0,0 +1,172 @@
+    type_: GraphQLTypeAnnotation<SelectableServerFieldId>,
+    indentation_level: u8,
+) -> String {
+    match type_ {
+        GraphQLTypeAnnotation::Named(named_inner_type) => {
+            format!(
+                "{} | null | void",
+                format_server_field_type(schema, named_inner_type.item, indentation_level)
+            )
+        }
+        GraphQLTypeAnnotation::List(list) => {
+            format!(
+                "ReadonlyArray<{}> | null",
+                format_server_field_type(schema, *list.inner(), indentation_level)
+            )
+        }
+        GraphQLTypeAnnotation::NonNull(non_null) => match *non_null {
+            GraphQLNonNullTypeAnnotation::Named(named_inner_type) => {
+                format_server_field_type(schema, named_inner_type.item, indentation_level)
+            }
+            GraphQLNonNullTypeAnnotation::List(list) => {
+                format!(
+                    "ReadonlyArray<{}>",
+                    format_server_field_type(schema, *list.inner(), indentation_level)
+                )
+            }
+        },
+    }
+}
+
+fn format_server_field_type(
+    schema: &ValidatedSchema,
+    field: SelectableServerFieldId,
+    indentation_level: u8,
+) -> String {
+    match field {
+        SelectableServerFieldId::Object(object_id) => {
+            let mut s = "{\n".to_string();
+            for (name, field_definition) in schema
+                .server_field_data
+                .object(object_id)
+                .encountered_fields
+                .iter()
+                .filter(|x| matches!(
+                    x.1,
+                    FieldDefinitionLocation::Server(server_field_id) if schema.server_field(*server_field_id).is_discriminator == false),
+                )
+            {
+                let field_type =
+                    format_field_definition(schema, name, field_definition, indentation_level + 1);
+                s.push_str(&field_type)
+            }
+            s.push_str(&format!("{}}}", "  ".repeat(indentation_level as usize)));
+            s
+        }
+        SelectableServerFieldId::Scalar(scalar_id) => schema
+            .server_field_data
+            .scalar(scalar_id)
+            .javascript_name
+            .to_string(),
+    }
+}
+
+fn format_field_definition(
+    schema: &ValidatedSchema,
+    name: &SelectableFieldName,
+    type_: &FieldDefinitionLocation<ServerFieldId, ClientFieldId>,
+    indentation_level: u8,
+) -> String {
+    match type_ {
+        FieldDefinitionLocation::Server(server_field_id) => {
+            let type_annotation = &schema.server_field(*server_field_id).associated_data;
+            format!(
+                "{}readonly {}: {},\n",
+                "  ".repeat(indentation_level as usize),
+                name,
+                format_type_annotation(schema, type_annotation, indentation_level + 1),
+            )
+        }
+        FieldDefinitionLocation::Client(_) => {
+            panic!(
```

Comment:
Ooooh interesting, I was thinking about this, and this might break 😂  There's nothing stopping you from defining a client field on an InputObject. LOL. Anyway, this is fine as it is, as people should not be doing that?

The reason you're allowed to is that Isograph doesn't have a distinction between input and output objects (IMO, that's silly and won't make sense in non-GraphQL worlds). So anyway, this panic is good

Reply:
#161